src/core/lib/event_engine/windows/grpc_polled_fd_windows.h

```diff
@@ -0,0 +1,210 @@
+    ares_ssize_t SendV(WSAErrorContext* wsa_error_ctx, const struct iovec* iov,
+                       int iov_count);
+    int Connect(WSAErrorContext* wsa_error_ctx, const struct sockaddr* target,
+                ares_socklen_t target_len);
+
+   private:
+    enum WriteState {
+      WRITE_IDLE,
+      WRITE_REQUESTED,
+      WRITE_PENDING,
+      WRITE_WAITING_FOR_VERIFICATION_UPON_RETRY,
+    };
+
+    void ScheduleAndNullReadClosure(absl::Status error);
+    void ScheduleAndNullWriteClosure(absl::Status error);
+    void ContinueRegisterForOnReadableLocked();
+    void ContinueRegisterForOnWriteableLocked();
+    int SendWriteBuf(LPDWORD bytes_sent_ptr, LPWSAOVERLAPPED overlapped,
+                     int* wsa_error_code);
+    ares_ssize_t SendVUDP(WSAErrorContext* wsa_error_ctx,
+                          const struct iovec* iov, int iov_count);
+    ares_ssize_t SendVTCP(WSAErrorContext* wsa_error_ctx,
+                          const struct iovec* iov, int iov_count);
+    void OnTcpConnect();
+    int ConnectUDP(WSAErrorContext* wsa_error_ctx,
+                   const struct sockaddr* target, ares_socklen_t target_len);
+    int ConnectTCP(WSAErrorContext* wsa_error_ctx,
+                   const struct sockaddr* target, ares_socklen_t target_len);
+    // TODO(apolcyn): improve this error handling to be less conversative.
+    // An e.g. ECONNRESET error here should result in errors when
+    // c-ares reads from this socket later, but it shouldn't necessarily cancel
+    // the entire resolution attempt. Doing so will allow the "inject broken
+    // nameserver list" test to pass on Windows.
+    void OnIocpReadable();
+    void OnIocpWriteable();
+
+    static grpc_slice FlattenIovec(const struct iovec* iov, int iov_count);
+
+    grpc_core::Mutex* mu_;
+    std::unique_ptr<WinSocket> winsocket_;
+    char recv_from_source_addr_[200];
+    ares_socklen_t recv_from_source_addr_len_;
+    grpc_slice read_buf_;
+    bool read_buf_has_data_ = false;
+    grpc_slice write_buf_;
+    absl::AnyInvocable<void(absl::Status)> read_closure_;
+    absl::AnyInvocable<void(absl::Status)> write_closure_;
+    std::unique_ptr<AnyInvocableClosure> outer_read_closure_;
+    std::unique_ptr<AnyInvocableClosure> outer_write_closure_;
+    const std::string name_;
+    bool shutdown_called_ = false;
+    int address_family_;
+    int socket_type_;
+    // State related to TCP sockets
+    AnyInvocableClosure on_tcp_connect_locked_;
+    bool connect_done_ = false;
+    int wsa_connect_error_ = 0;
+    WriteState tcp_write_state_ = WRITE_IDLE;
+    // We don't run register_for_{readable,writeable} logic until
+    // a socket is connected. In the interim, we queue readable/writeable
+    // registrations with the following state.
+    bool pending_continue_register_for_on_readable_locked_ = false;
+    bool pending_continue_register_for_on_writeable_locked_ = false;
+    EventEngine* event_engine_;
```

Comment:
Good point! This is interesting. `GrpcPolledFdWindows` has a shared ownership model and is owned by the `AresResolver`, the `GrpcPolledFdFactory` and the pending IOCP callbacks. The `AresResolver` owns the `GrpcPolledFdFactory`. And it seems like `GrpcPolledFdWindows` could outlive the `AresResolver`, since it could be owned by the pending IOCP callbacks. But the pending read/write callback takes a Ref from the `AresResolver` and would prevent it from being destroyed. This `EventEngine*` is initialized from the stored pointer inside the `std::shared_ptr<EventEngine>` owned by the `AresResolver` and should be valid at the time of use. Added the comment.